hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala

```diff
@@ -482,11 +514,40 @@ extends Importer:
                   case ((pss, ctx), ps) => 
                     val (qs, newCtx) = params(ps)(using ctx)
                     (pss :+ ParamList(ParamListFlags.empty, qs), newCtx)
+              // * Elaborate signature
+              val s = td.signature.orElse(newSignatureTrees.get(id.name)).map(term)
               val b = rhs.map(term(_)(using newCtx))
               val r = FlowSymbol(s"‹result of ${sym}›", nextUid)
-              val tdf = TermDefinition(owner, k, sym, pss,
-                td.signature.orElse(newSignatureTrees.get(id.name)).map(term), b, r)
+              val tdf = TermDefinition(owner, k, sym, pss, s, b, r, TermDefFlags(mod))
               sym.defn = S(tdf)
+
+              // the return type of the function
+              val result = td.head match
+                case InfixApp(_, Keyword.`:`, rhs) => S(term(rhs)(using newCtx))
```

Comment:
What. I thought that's the problem we've just discussed earlier.

In this https://github.com/hkust-taco/mlscript/pull/233#discussion_r1842335992 you said "signatures are only picked up from externally-given declaration"; and at

https://github.com/hkust-taco/mlscript/blob/e3859934e2c38db20f2a6f6998aba55dc49940a9/hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala#L234

it does not always pick up the result type from the parsed tree. That's why I did another parsing to make sure result types are always parsed.

Reply:
Ok, let me be completely explicit, so we don't get hung up on details and terminology:

* There should be only one place that determines what is the inline-annotated result type in a definition (if there is any). The logic for finding it by analyzing trees should not be duplicated (as you are doing now), as it means that any update to it must also be duplicated everywhere, which is impractical and error-prone.
  Currently, this is done in `TypeOrTermDef` and the result is confusingly called `signature` (we could change it to `annotatedResultType`).
* In the elaborator, if we see there is no inline-annotated result type, we should use the one given by the separate signature. This could be implemented later, though.
* Such explicitly-annotated result types must not be elaborated several times. Indeed, currently that's not the case, but we should be careful that this does not happen whenever we add the logic to achieve the previous point.

> it does not always pick up the result type from the parsed tree. That's why I did another parsing to make sure result types are always parsed.

If the implementation in `TypeOrTermDef` is missing some cases, you should fix that implementation, instead of leaving it broken and reimplementing it separately!